### ISSUE TITLE
Precise error codes when using the SFTP subsystem

### ISSUE THREAD

**Issue by koutheir** [alexcrichton/ssh2-rs#182]
When an error happens while using the SFTP subsystem, the error returned by `libssh2` is `LIBSSH2_ERROR_SFTP_PROTOCOL`, which only indicate the origin of the error, but not its nature. In order to know what caused the error, `libssh2_sftp_last_error()` needs to be called, and would return one of the `LIBSSH2_FX_*` error codes.

The `libssh2` crate doesn't call `libssh2_sftp_last_error()` when `LIBSSH2_ERROR_SFTP_PROTOCOL` is reported, but simply returns an obscure `Error`.

The `libssh2_sys` crate already provides `libssh2_sftp_last_error()` but there is no way to call it without having access to the `LIBSSH2_SFTP` handle from the `libssh2` crate (no public function allows that).

Is there any other way I could do this?

**Comment by wez** [comment]
I don't think you can do much about this without making some changes in the rust crate to expose the error code.  I'd happily review a PR for that; my suggestion after only a brief glance at the code is:

* Add a new `SftpErrorCode` error type to represent the differently namespaced error codes
* Add an `SftpError` enum that can be either `SftpErrorCode` or the main `Error` code.  I'm honestly flexible on the Error stuff: use your judgement and prefer something that feels ergonomic + useful if you think of something nicer.
* Add a helper function to sftp.rs that accepts both the session and the sftp pointer and that returns `SftpError`.
* Adopt the use of that helper where appropriate in `sftp.rs`; look for calls to `sess.rc` for the first candidates that might need changing.

**Comment by koutheir** [comment]
> Add a new SftpErrorCode error type to represent the differently namespaced error codes

I went deeper in the structure. I replaced the `c_int` error code by an enumeration that separates the two error name spaces (Session error codes, and SFTP error codes). I kept using one `Error`.

> Add a helper function to sftp.rs that accepts both the session and the sftp pointer and that returns SftpError.

I added a few helper functions to `Sftp` and `File` in order to make sure error management is uniform and performed in a single place.

Further changes in `libssh2-sys`:

- Added declaration of the method `libssh2_session_last_errno()`.
- Corrected the types of SFTP error constants in `libssh2-sys` from `c_int` to `c_ulong` because that is what `libssh2_sftp_last_error()` returns.

Further changes in `libssh2`:
- Changed some methods of the `Error` type to make clear distinction between session errors and SFTP errors.

Please confirm that these changes make sense to you. In the meantime, I'll make a pull request.

**Comment by harmic** [comment]
I'm also waiting on this - any help needed to progress it?
@koutheir's PR fails checks, but they seem to relate to building a dependency `ext-prim` which has not yet been updated to use `llvm-asm!` instead of `asm!`, so they don't seem related to the PR.

**Comment by koutheir** [comment]
I'm also waiting...

**Comment by crisidev** [comment]
New PR implementing this based on latest version of libssh2-rs: https://github.com/alexcrichton/ssh2-rs/pull/203

**Comment by crisidev** [comment]
Resolving as the PR is merged and version 0.9.0 has been released.
